File: source/c#/Form1.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.IO.Ports;
using System.Windows.Forms;


namespace USB_Switch
{
	public partial class Form1 : Form
	{
		public Form1()
		{
			this.InitializeComponent();
		}

		private void Form1_Load(object sender, EventArgs e)
		{
			foreach (string item in SerialPort.GetPortNames())
			{
				this.comBox_COM.Items.Add(item);
			}
		}

		private void btn_connect_Click(object sender, EventArgs e)
		{
			if (this.sp != null)
			{
				try
				{
					this.sp.DataReceived -= this.SpDataReceived;
					this.sp.Dispose();
					this.sp.Close();
					this.sp = null;
					this.btn_connect.Text = "连接";
					this.btn_close_n.Enabled = false;
					this.btn_open_n.Enabled = false;
					this.btn_flash.Enabled = false;
					this.button1.Enabled = false;
					this.button2.Enabled = false;
					this.button3.Enabled = false;
					this.button4.Enabled = false;
					this.button5.Enabled = false;
					this.button6.Enabled = false;
					this.button7.Enabled = false;
					this.button8.Enabled = false;
					this.button9.Enabled = false;
					this.button10.Enabled = false;
					this.button11.Enabled = false;
					this.button12.Enabled = false;
					this.button13.Enabled = false;
					this.button14.Enabled = false;
					this.button15.Enabled = false;
					this.button16.Enabled = false;
					this.button17.Enabled = false;
					this.button18.Enabled = false;
					this.button19.Enabled = false;
					this.button20.Enabled = false;
					this.button21.Enabled = false;
				}
				catch (Exception ex)
				{
					this.sp = null;
					MessageBox.Show(ex.ToString());
				}
			}
			else if (this.comBox_COM.Text == "请选择串口")
			{
				MessageBox.Show("请选择正确的串口");
			}
			else
			{
				try
				{
					this.sp = new SerialPort(this.comBox_COM.Text, 0x2580);
					this.sp.DataBits = 0x8;
					this.sp.Parity = Parity.None;
					this.sp.StopBits = StopBits.One;
					this.sp.Open();
					this.sp.DataReceived += this.SpDataReceived;
					this.btn_connect.Text = "断开";
					this.btn_close_n.Enabled = true;
					this.btn_open_n.Enabled = true;
					this.btn_flash.Enabled = true;
					this.button1.Enabled = true;
					this.button2.Enabled = true;
					this.button3.Enabled = true;
					this.button4.Enabled = true;
					this.button5.Enabled = true;
					this.button6.Enabled = true;
					this.button7.Enabled = true;
					this.button8.Enabled = true;
					this.button9.Enabled = true;
					this.button10.Enabled = true;
					this.button11.Enabled = true;
					this.button12.Enabled = true;
					this.button13.Enabled = true;
					this.button14.Enabled = true;
					this.button15.Enabled = true;
					this.button16.Enabled = true;
					this.button17.Enabled = true;
					this.button18.Enabled = true;
					this.button19.Enabled = true;
					this.button20.Enabled = true;
					this.button21.Enabled = true;
				}
				catch (Exception ex)
				{
					this.sp = null;
					MessageBox.Show(ex.ToString());
				}
			}
		}

		private void SpDataReceived(object sender, SerialDataReceivedEventArgs e)
		{
			int bytesToRead = this.sp.BytesToRead;
			byte[] buf = new byte[bytesToRead];
			this.sp.Read(buf, 0x0, bytesToRead);
			base.Invoke(new EventHandler(delegate(object param0, EventArgs param1)
			{
				for (int i = 0x0; i < buf.Length; i++)
				{
					TextBoxBase textBoxBase = this.txt_back;
					int num = (int)buf[i];
					textBoxBase.AppendText(num.ToString("X2") + " ");
				}
			}));
		}

		private void btn_open_n_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[] { 0xA0, 0x1, 0x1, 0xA2 };
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void btn_close_n_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x1,
					0x0,
					0xA1
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button19_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x5,
					0x2,
					0xA7
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button18_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x6,
					0x1,
					0xA7
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button17_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x6,
					0x0,
					0xA6
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button16_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x6,
					0x2,
					0xA8
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button15_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x7,
					0x1,
					0xA8
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button9_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x2,
					0x2,
					0xA4
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button8_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x2,
					0x0,
					0xA2
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button7_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x2,
					0x1,
					0xA3
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button6_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x3,
					0x1,
					0xA4
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button5_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x3,
					0x0,
					0xA3
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button4_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x3,
					0x2,
					0xA5
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button2_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x4,
					0x0,
					0xA4
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button14_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x7,
					0x0,
					0xA7
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button13_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x7,
					0x2,
					0xA9
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button12_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x0,
					0x1,
					0xA1
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button11_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x0,
					0x0,
					0xA0
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button10_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x0,
					0x2,
					0xA2
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button3_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x4,
					0x1,
					0xA5
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button1_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x4,
					0x2,
					0xA6
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void btn_close_y_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x1,
					0x2,
					0xA3
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button20_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x5,
					0x0,
					0xA5
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		private void button21_Click(object sender, EventArgs e)
		{
			this.txt_back.Text = "";
			try
			{
				byte[] buffer = new byte[]
				{
					0xA0,
					0x5,
					0x1,
					0xA6
				};
				this.sp.Write(buffer, 0x0, 0x4);
			}
			catch (Exception ex)
			{
				MessageBox.Show(ex.ToString());
			}
		}

		public SerialPort sp;
	}
}

File: source/VB.net/Switch_VB/Form1.vb
Imports System.IO.Ports


Partial Public Class Form1
    Inherits Form

    Public wport As SerialPort

    Public Sub New()
        Me.InitializeComponent()
    End Sub

    Private Sub btn_close_n_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_close_n.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 1, 0, 161}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub btn_close_y_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_flash.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 1, 2, 163}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub btn_connect_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_connect.Click
        If Me.wport IsNot Nothing Then

            Try
                RemoveHandler Me.wport.DataReceived, AddressOf Me.wport_DataReceived
                Me.wport.Dispose()
                Me.wport.Close()
                Me.wport = Nothing
                Me.btn_connect.Text = "连接"
                Me.btn_close_n.Enabled = False
                Me.btn_open_n.Enabled = False
                Me.btn_flash.Enabled = False
                Me.button1.Enabled = False
                Me.button2.Enabled = False
                Me.button3.Enabled = False
                Me.button4.Enabled = False
                Me.button5.Enabled = False
                Me.button6.Enabled = False
                Me.button7.Enabled = False
                Me.button8.Enabled = False
                Me.button9.Enabled = False
                Me.button10.Enabled = False
                Me.button11.Enabled = False
                Me.button12.Enabled = False
                Me.button13.Enabled = False
                Me.button14.Enabled = False
                Me.button15.Enabled = False
                Me.button16.Enabled = False
                Me.button17.Enabled = False
                Me.button18.Enabled = False
                Me.button19.Enabled = False
                Me.button20.Enabled = False
                Me.button21.Enabled = False
            Catch ex As Exception
                Me.wport = Nothing
                MessageBox.Show(ex.ToString())
            End Try
        ElseIf Me.comBox_COM.Text = "请选择串口" Then
            MessageBox.Show("请选择正确的串口")
        Else

            Try
                Me.wport = New SerialPort(Me.comBox_COM.Text, 921600)
                Me.wport.DataBits = 8
                Me.wport.Parity = Parity.None
                Me.wport.StopBits = StopBits.One
                Me.wport.Open()
                AddHandler Me.wport.DataReceived, AddressOf Me.wport_DataReceived
                Me.btn_connect.Text = "断开"
                Me.btn_close_n.Enabled = True
                Me.btn_open_n.Enabled = True
                Me.btn_flash.Enabled = True
                Me.button1.Enabled = True
                Me.button2.Enabled = True
                Me.button3.Enabled = True
                Me.button4.Enabled = True
                Me.button5.Enabled = True
                Me.button6.Enabled = True
                Me.button7.Enabled = True
                Me.button8.Enabled = True
                Me.button9.Enabled = True
                Me.button10.Enabled = True
                Me.button11.Enabled = True
                Me.button12.Enabled = True
                Me.button13.Enabled = True
                Me.button14.Enabled = True
                Me.button15.Enabled = True
                Me.button16.Enabled = True
                Me.button17.Enabled = True
                Me.button18.Enabled = True
                Me.button19.Enabled = True
                Me.button20.Enabled = True
                Me.button21.Enabled = True
            Catch ex As Exception
                Me.wport = Nothing
                MessageBox.Show(ex.ToString())
            End Try
        End If
    End Sub

    Private Function btn_connect_Click() As EventHandler
        Throw New NotImplementedException
    End Function

    Private Sub btn_open_n_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_open_n.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 1, 1, 162}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 4, 2, 166}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button10_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button10.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 0, 2, 162}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button11_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button11.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 0, 0, 160}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button12_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button12.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 0, 1, 161}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button13_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button13.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 7, 2, 169}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button14_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button14.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 7, 0, 167}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button15_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button15.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 7, 1, 168}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button16_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button16.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 6, 2, 168}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button17_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button17.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 6, 0, 166}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button18_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button18.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 6, 1, 167}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button19_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button19.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 5, 2, 167}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button2.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 4, 0, 164}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button20_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button20.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 5, 0, 165}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button21_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button21.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 5, 1, 166}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button3_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button3.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 4, 1, 165}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button4_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button4.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 3, 2, 165}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button5_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button5.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 3, 0, 163}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button6_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button6.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 3, 1, 164}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button7_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button7.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 2, 1, 163}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button8_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button8.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 2, 0, 162}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub button9_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button9.Click
        Me.txt_back.Text = ""

        Try
            Dim buffer As Byte() = New Byte() {160, 2, 2, 164}
            Me.wport.Write(buffer, 0, 4)
        Catch ex As Exception
            MessageBox.Show(ex.ToString())
        End Try
    End Sub

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        For Each item As String In SerialPort.GetPortNames()
            Me.comBox_COM.Items.Add(item)
        Next
    End Sub
    Private Sub wport_DataReceived(ByVal sender As Object, ByVal e As SerialDataReceivedEventArgs)
        Dim bytesToRead As Integer = Me.wport.BytesToRead
        Dim buf As Byte() = New Byte(bytesToRead - 1) {}
        Me.wport.Read(buf, 0, bytesToRead)
        MyBase.Invoke(New EventHandler(Sub(ByVal param0 As Object, ByVal param1 As EventArgs)

                                           For i As Integer = 0 To buf.Length - 1
                                               Dim textBoxBase As TextBoxBase = Me.txt_back
                                               Dim num As Integer = CInt(buf(i))
                                               textBoxBase.AppendText(num.ToString("X2") & " ")
                                           Next
                                       End Sub))
    End Sub
End Class